torch/distributed/pipelining/schedules.py

```diff
@@ -2735,6 +2736,257 @@ def _calculate_single_rank_operations(self, rank) -> list[Optional[_Action]]:
+                    ),
+                )
+
+                # Update counters for sub_actions
+                counters[forward_key] = forward_mb + 1
+                counters[backward_key] = backward_mb + 1
+
+                # Handle FULL_BACKWARD counter updates
+                input_key = (backward_stage, _ComputationType.BACKWARD_INPUT)
+                weight_key = (backward_stage, _ComputationType.BACKWARD_WEIGHT)
+                counters[input_key] = counters.get(input_key, 0) + 1
+                counters[weight_key] = counters.get(weight_key, 0) + 1
+
+                # Add the overlapped action
+                actions.append(
+                    _Action(-1, _ComputationType.OVERLAP_F_B, None, sub_actions)
+                )
+            else:
+                # Regular single action
+                key = (stage_index, computation_type)
+                mb_index = counters.get(key, 0)
+                actions.append(_Action(stage_index, computation_type, mb_index))
+                counters[key] = mb_index + 1
+
+                # If FULL_BACKWARD is used, also increment the separate BACKWARD_INPUT and BACKWARD_WEIGHT counters
+                # so they skip this microbatch since it's already covered by FULL_BACKWARD
+                if computation_type == _ComputationType.FULL_BACKWARD:
+                    input_key = (stage_index, _ComputationType.BACKWARD_INPUT)
+                    weight_key = (stage_index, _ComputationType.BACKWARD_WEIGHT)
+                    counters[input_key] = counters.get(input_key, 0) + 1
+                    counters[weight_key] = counters.get(weight_key, 0) + 1
+
+                # If BACKWARD_INPUT is updated, add corresponding weight action to queue and increment FULL_BACKWARD counter
+                if computation_type == _ComputationType.BACKWARD_INPUT:
+                    # Add weight action to queue for later processing
+                    weight_queue.append((stage_index, mb_index))
+                    full_backward_key = (stage_index, _ComputationType.FULL_BACKWARD)
+                    counters[full_backward_key] = counters.get(full_backward_key, 0) + 1
+
+        def add_weight_action(actions: list, counters: dict):
+            """Helper method to add a weight action from the queue."""
+            if not weight_queue:
+                return  # No pending weight actions, skip
+            # Pop the oldest weight action from the queue
+            actual_stage_index, weight_mb_index = weight_queue.pop(0)
+            actions.append(
+                _Action(
+                    actual_stage_index,
+                    _ComputationType.BACKWARD_WEIGHT,
+                    weight_mb_index,
+                )
+            )
+            # Update the counter for the actual stage that was processed
+            weight_key = (actual_stage_index, _ComputationType.BACKWARD_WEIGHT)
+            counters[weight_key] = counters.get(weight_key, 0) + 1
+
+        # Step 1: F0
```

Comment:
generally, is there some matching between the 6 steps and the rank-based formulas you used here, and the paper/code?  it isn't that obvious how this compares to the logic in [dualpipev](https://github.com/deepseek-ai/DualPipe/blob/main/dualpipe/dualpipev.py#L120)

Reply:
The logic for dualpipev is here (https://github.com/deepseek-ai/DualPipe/blob/3da1bbea53606543d7f5f232338fc58096db30e3/dualpipe/dualpipev.py#L331-L396). So this schedule copies that (minus their communication operations, which we put in ourselves).

The overlapped_f_b in the dualpipev is a dummy implementation using an MLP. The DeepSeek team does not provide a concrete implementation for what they described in the paper, but they leverage the tech built in [DeepEP](https://github.com/deepseek-ai/DeepEP/blob/main/deep_ep/buffer.py#L13-L27) for the dispatch + MOE + combine example (https://github.com/deepseek-ai/DeepEP/blob/main/README.md#example-use-in-model-training-or-inference-prefilling)